packages/adapters/fastify/web/src/web.ts

```diff
@@ -0,0 +1,179 @@
+
+import { getConfig, getPaths } from '@redwoodjs/project-config'
+
+export interface RedwoodFastifyWebOptions {
+  redwood?: {
+    apiUrl?: string
+    apiUpstreamUrl?: string
+
+    /**
+     * @deprecated Use `apiUpstreamUrl` instead.
+     */
+    apiHost?: string
+  }
+}
+
+export async function redwoodFastifyWeb(
+  fastify: FastifyInstance,
+  opts: RedwoodFastifyWebOptions,
+  done: HookHandlerDoneFunction
+) {
+  fastify.log.debug('opts')
+  fastify.log.debug(opts)
+
+  const options = resolveOptions(opts)
+
+  fastify.log.debug('resolving options')
+  fastify.log.debug(options)
+
+  await fastify.register(fastifyUrlData)
+
+  // Serve prerendered files directly, instead of the index
+  const prerenderedFiles = await fg('**/*.html', {
+    cwd: getPaths().web.dist,
+    ignore: ['index.html', '200.html', '404.html'],
+  })
+
+  for (const prerenderedFile of prerenderedFiles) {
+    const [pathName] = prerenderedFile.split('.html')
+
+    fastify.get(`/${pathName}`, (_, reply) => {
+      reply.header('Content-Type', 'text/html; charset=UTF-8')
+      reply.sendFile(prerenderedFile)
+    })
+  }
+
+  fastify.log.debug(`registering static assets at ${getPaths().web.dist}`)
+  fastify.log.debug(getPaths().web.dist)
+  // Serve static assets
+  fastify.register(fastifyStatic, {
+    root: getPaths().web.dist,
+  })
+
+  // If `apiUpstreamUrl` is set, proxy requests from `apiUrl` to `apiUpstreamUrl`.
+  // In this case, `apiUrl` has to be relative; `resolveOptions` above throws if it's not
+  if (options.redwood.apiUpstreamUrl) {
+    fastify.log.debug('registering proxy')
+
+    fastify.register(httpProxy, {
+      prefix: options.redwood.apiUrl,
+      upstream: options.redwood.apiUpstreamUrl,
+      disableCache: true,
+    })
+  }
+
+  // If `200.html` exists, the project has been prerendered.
+  // If it doesn't, fallback to the default (`index.html`)
+  const prerenderIndexPath = path.join(getPaths().web.dist, '200.html')
+
+  const fallbackIndexPath = fs.existsSync(prerenderIndexPath)
+    ? '200.html'
+    : 'index.html'
+
+  // For SPA routing, fallback on unmatched routes and let client-side routing take over
+  fastify.setNotFoundHandler({}, (req, reply) => {
+    const urlData = req.urlData()
+    const requestedExtension = path.extname(urlData.path ?? '')
+
+    // Paths with no extension (`/about`) or an .html extension (`/about.html`)
+    // should be handled by the client side router.
+    // See the discussion in https://github.com/redwoodjs/redwood/pull/9272.
+    if (requestedExtension === '' || requestedExtension === '.html') {
+      reply.header('Content-Type', 'text/html; charset=UTF-8')
+      return reply.sendFile(fallbackIndexPath)
+    }
+
+    // If we got here, the user is requesting an asset with an extension
+    // (like `profile.png`) that doesn't exist
+    reply.code(404)
+    return reply.send('Not Found')
+  })
+
+  done()
+}
+
+export function resolveOptions(options: RedwoodFastifyWebOptions) {
```

Comment:
I'd move this function to another file. That way it feels more natural to export it. Which also keeps it testable.
And if we do that, web.ts can be used instead of index.ts in this package. So we could completely get rid of index.ts

Added benefit, which we might not need (yet), is that it can then be mocked when testing `redwoodFastifyWeb`

Reply:
Sounds good; done in https://github.com/redwoodjs/redwood/pull/9884/commits/e477c530410bac2703a241b8de7bd1d12b430d09